libraries/ESP8266WiFi/src/WiFiUdp.cpp

```diff
@@ -176,6 +176,24 @@ int WiFiUDP::endPacket()
     return (_ctx->send()) ? 1 : 0;
 }
 
+bool WiFiUDP::endPacketMayRetry()
```

Comment:
As I tried to explain in our internal discussion, there is an interaction between this new method and scheduled functions, if both try to use the same Udp object. The problem happens because this new method blocks while waiting for the pbuf to progress enough to make room, and during that time a scheduled function could get triggered and try to beginPacket().
As you pointed out, this is a reentrancy case, and the current Udp implementation doesn't support reentrancy. However, the case can be hit easily by a user, as I try to show in the gist [there](https://gist.github.com/devyte/2c7c5404294f93c20c841179e88291cd). The example is trivial, and it is not clear why it doesn't work as expected.
To mitigate the problem, I think the beginPacket() method should detect whether there is data pending in the UdpContext tx buffer, and if so error out. That can mean either return false (up to the user to check the return value of beginPacket()), or fail an assert (programming error, so crater on purpose with some debug message).
I'm not sure if the detection can be done within beginPaket() itself. but if not, I think it would be valid to detect inside UdpContext::connect(), something along the lines of:
```cpp
if(tx_buffer_has_data())
  return false;
```
at the top.